ddsrecorder_participants/src/cpp/mcap/McapHandler.cpp

```diff
@@ -0,0 +1,460 @@
+        const Message& msg)
+{
+    std::lock_guard<std::mutex> guard(write_mtx_);
+
+    samples_buffer_.push(msg);
+    if (!paused_ && samples_buffer_.size() == buffer_size_)
+    {
+        logInfo(DDSRECORDER_MCAP_HANDLER, "Full buffer, writting to disk...");
+        dump_data_nts_();
+    }
+}
+
+void McapHandler::add_pending_samples_(
+        const std::string& schema_name)
+{
+    assert(pending_samples_.find(schema_name) != pending_samples_.end());
+    auto& pending_queue = pending_samples_[schema_name];
+
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Sending pending samples of type: " << schema_name << ".");
+
+    while (!pending_queue.empty())
+    {
+        auto& sample = pending_queue.front();
+        DdsTopic sample_topic;
+        sample_topic.m_topic_name = sample.first;
+        sample_topic.type_name = schema_name;
+        auto channel_id = get_channel_id_(sample_topic);
+        auto& msg = sample.second;
+        msg.channelId = channel_id;
+        try
+        {
+            add_data_(msg);
+        }
+        catch(const utils::Exception& e)
+        {
+            throw utils::InconsistencyException(
+                STR_ENTRY << "Error writting in MCAP a message in topic " << sample.first
+            );
+        }
+        pending_queue.pop();
+    }
+}
+
+void McapHandler::event_thread_routine_()
+{
+    while (paused_)
+    {
+        clear_all_();
+
+        bool timeout;
+        {
+            std::unique_lock<std::mutex> lock(event_cv_mutex_);
+
+            timeout = !event_cv_.wait_for(
+                lock,
+                std::chrono::seconds(event_window_),
+                [&]
+                {
+                    return event_triggered_;
+                });
+            event_triggered_ = false;
+        }
+
+        if (timeout)
+        {
+            logInfo(DDSRECORDER_MCAP_HANDLER, "Event thread timeout.");
+        }
+        else
+        {
+            if (paused_)
+            {
+                logInfo(DDSRECORDER_MCAP_HANDLER, "Event triggered: dumping buffered data.");
+                dump_data_();
+
+            }
+            else
+            {
+                logInfo(DDSRECORDER_MCAP_HANDLER, "Finishing event thread routine.");
+                clear_all_();
+                break;
+            }
+        }
+    }
+}
+
+void McapHandler::stop_event_thread_()
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Stopping event thread.");
+    paused_.store(false);
+    if (event_thread_.joinable())
+    {
+        {
+            std::lock_guard<std::mutex> lock(event_cv_mutex_);
+            event_triggered_ = true;
+        }
+        event_cv_.notify_one();
+        event_thread_.join();
+    }
+}
+
+void McapHandler::clear_all_()
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Cleaning all buffers.");
+
+    {
+        // Clear samples buffer
+        std::lock_guard<std::mutex> guard(write_mtx_);
+        std::queue<Message> empty;
+        std::swap(samples_buffer_, empty);
```

Comment:
Is swap the correct way? I saw stack overflow, I just don't think is very readable to have a swap here.